### PR TITLE
improve worst-case performance of BTreeSet intersection v3

### PR DESCRIPTION
**Pull request by ssomers**
Variation of [#59078](https://github.com/rust-lang/rust/pull/59078) with `Intersection` remaining a struct

r? @scottmcm

### CODE REVIEW

**Comment by KodrAus** [comment]
Hi @ssomers :wave:

Would you prefer to close #58577 and #59078 and look at this PR instead?

I haven't had a chance to look at what's changed since the changeset I approved, could you summarize the differences? It looks like this one does keep the `enum` for the strategy private which I think is an improvement on the v2 implementation.

**Comment by ssomers** [comment]
Sure. The difference with the approved state is:
- struct Intersection has **both fields replaced** by the enum, even though both implementations share the same iterator. I think it makes the code more readable and it simplifies changing/adding algorithms that don't need that field.
- The gut of both implementations is exposed (but hidden, again) as **unstable feature**, to allow the benchmark to compare them.
- Thus new benchmarks contrast the performance of both implementations with the actual implementation chosen by the size rule.
- The size rule favors the new implementation much more than before, but still less than looks optimal (on my system, for this benchmark, with these kind of elements). And if one of the sets is empty, the other set's iterator isn't even constructed.
- Sets are always ordered by size, because it simplifies code and because size_hint already had to choose the small set.
- The classic implementation no longer relies on Peekable (which greatly helps for union and difference, but not here).

File: src/liballoc/collections/btree/set.rs [hunk 1]
```diff
@@ -326,9 +342,51 @@ impl<T: Ord> BTreeSet<T> {
     /// ```
     #[stable(feature = "rust1", since = "1.0.0")]
     pub fn intersection<'a>(&'a self, other: &'a BTreeSet<T>) -> Intersection<'a, T> {
+        let (a_set, b_set) = if self.len() <= other.len() {
+            (self, other)
+        } else {
+            (other, self)
+        };
+        if a_set.len() > b_set.len() / 16 {
```
Comment (KodrAus) [inline src/liballoc/collections/btree/set.rs]: Might be worth leaving a comment here about what this branch is for and why we use the constant `16` to decide which strategy to use.

File: src/liballoc/collections/btree/set.rs [hunk 2]
```diff
@@ -1083,24 +1149,35 @@ impl<'a, T: Ord> Iterator for Intersection<'a, T> {
...
+                        return Some(a_next);
+                    }
                 }
             }
         }
     }
 
     fn size_hint(&self) -> (usize, Option<usize>) {
-        (0, Some(min(self.a.len(), self.b.len())))
+        let max_size = match &self.inner {
```
Comment (KodrAus) [inline src/liballoc/collections/btree/set.rs]: This would actually be `min_size`, right?

File: src/liballoc/collections/btree/set.rs [hunk 3]
```diff
@@ -164,17 +164,33 @@ impl<T: fmt::Debug> fmt::Debug for SymmetricDifference<'_, T> {
...
+    inner: IntersectionInner<'a, T>,
+}
+#[derive(Debug)]
+enum IntersectionInner<'a, T: 'a> {
+    Stitch {
+        a_iter: Iter<'a, T>, // for size_hint, should be the smaller of the sets
+        b_iter: Iter<'a, T>,
+    },
+    Search {
+        a_iter: Iter<'a, T>, // for size_hint, should be the smaller of the sets
```
Comment (KodrAus) [inline src/liballoc/collections/btree/set.rs]: Since the implementation depends on `a_iter` being smaller I think we should choose more descriptive names here for `a` and `b`.

File: src/liballoc/benches/lib.rs [hunk 4]
```diff
@@ -1,5 +1,6 @@
 #![feature(repr_simd)]
 #![feature(test)]
+#![feature(benches_btree_set)]
```
Comment (KodrAus) [inline src/liballoc/benches/lib.rs]: Before merging I think we should remove this feature and make the methods private (or just inline them). It's nice to get an idea of the performance characteristics of each strategy, but once we understand those I think we can move forward with just the general benchmarks.

File: src/liballoc/collections/btree/set.rs (hunk 2 above)
Comment (ssomers) [reply src/liballoc/collections/btree/set.rs]: Had I taken the time to look up the doc of size_hint, I would have called it `upper_bound`. It's the "min" of the input sets, and "max" of the size_hint, so I'm not a fan of `min_size' either. How about `min_len`?

File: src/liballoc/collections/btree/set.rs (hunk 3 above)
Comment (ssomers) [reply src/liballoc/collections/btree/set.rs]: Frankly, I hate the name `a` because it's intermingled with the same lifetime identifier, but didn't dare to change them. I'll go for `small` and `large` in the Search case, `small` and `other` in the Stitch case, unless you object.

File: src/liballoc/benches/lib.rs (hunk 4 above)
Comment (ssomers) [reply src/liballoc/benches/lib.rs]: But we only understand the performance characteristics:

- on 1 platform and 1 machine (I could build it on Linux on another machine though, but nothing modern),
- for one generation of the BTreeSet implementation. I see significant improvement from stable to nightly in my BTreeSet macro-benchmarks, tens of percents, with the same intersection implementation.

What alternatives are there for the feature litter?
- Moving the benchmarks closer to the lib code is terrible: every tweak or comment requires over an hour to build.
- Duplicating the lib code near/in the benchmark file is doable. You should notice if it's not up to date anymore when you check that the actual performance matches that of either implementation.
- Or similarly, duplicating the lib code and benchmarks in some separate repository (like I already did in https://github.com/ssomers/Bron-Kerbosch/blob/master/rust/bron_kerbosch/src/util.rs).

> (or just inline them)

I don't understand. Are you saying that Rust has ways to get around the privateness of the enum, without exposing _something_ as public?

**Comment by ssomers** [comment]
I made the benchmarks more fine-grained with set sizes and [plotted the results](https://plot.ly/~stein.somers/216/) (still only one machine). I think this highly suggests that the ideal strategy rule is much more complicated (not just logarithmic with size), and if that rule doesn't spend the performance it gains on evaluating itself, it could be greatly off on a system with different word size, cache sizes and architecture.

But it also confirms that factor 16 seems quite reasonable. To prevent the <30% performance hit for intersection of particularly crafted large sets, it would have to be 19. From the viewpoint of random sets, factor 16 should be lowered instead.

But if I commit all 146 benchmarks I have now, it's simply annoying for someone casually checking performance in general. So I'm becoming convinced that it's better to move this case study over to a separate repository, and leave in this PR only the final rule, and the few benchmarks already merged in earlier (or less), and thus nothing exposed as public unstable. Checking the rust source code, there are other references to github repositories besides rust-lang so I can probably just create one myself.

File: src/liballoc/benches/lib.rs (hunk 4 above)
Comment (KodrAus) [reply src/liballoc/benches/lib.rs]: @ssomers I think duplicating the code in a separate repository is the best way to go here :+1: I _believe_ we can run benchmarks in CI here (I don't remember the command off the top of my head) so we can worry about this later.

**Review by ssomers — COMMENTED** [review]
Here's the new comment (cause it seems rather difficult to find in the previous comment)

Oh well, this seems worse and can't be deleted.

File: src/liballoc/collections/btree/set.rs (hunk 1 above)
Comment (ssomers) [reply src/liballoc/collections/btree/set.rs]: Comment left, but it seems you have to find it yourself.

**Comment by ssomers** [comment]
I configured Travis on [the separate repository](https://github.com/ssomers/rust_bench_btreeset_intersection) and the readings on the linux build on that (virtual) machine are quite similar (with much less stability, as one would expect). The ideal factor is < 1 higher accross the range of sizes.

File: src/liballoc/collections/btree/set.rs (hunk 2 above)
Comment (KodrAus) [reply src/liballoc/collections/btree/set.rs]: `min_len` is ok with me :+1:

File: src/liballoc/collections/btree/set.rs (hunk 3 above)
Comment (KodrAus) [reply src/liballoc/collections/btree/set.rs]: That sounds good!

**Comment by KodrAus** [comment]
Thanks for all your investigation work @ssomers!

Just for good measure, let's add another test case to `liballoc::tests::btree::set::test_intersection` that checks intersection of a very large set with a very small one so we can be sure it'll be using the other strategy. Something like:

```rust
check_intersection(&[11, 5000, 1, 3, 77, 8924, 103],
    &(0..1000).collect::<Vec<_>>(),
    &[11, 1, 3, 77, 103]);
```

**Comment by ssomers** [comment]
I thought the test cases [in the first PR](https://github.com/rust-lang/rust/pull/58577/files) were still there and they didn't even include non-subset samples. Rest assured, the proptest in the separate repository covered everything.

So are we there now? Nope, I just realized that is_subset/superset is really quite similar. I'll keep that out of this PR, but it might mean that the 16 becomes a named constant.

**Comment by ssomers** [comment]
I cooked up a similar change to the implementation of set difference. It only needs half of the peekables, and it benefits from the same performance boost as with intersection if the right hand set is huge. I cooked up similar code resulting in:

before:

    test btree::set::difference_random_100_vs_100            ... bench:         914 ns/iter (+/- 27)
    test btree::set::difference_random_100_vs_10k            ... bench:      45,641 ns/iter (+/- 605)
    test btree::set::difference_random_10k_vs_100            ... bench:      63,561 ns/iter (+/- 512)
    test btree::set::difference_random_10k_vs_10k            ... bench:     205,747 ns/iter (+/- 5,792)
    test btree::set::difference_staggered_100_vs_100         ... bench:         989 ns/iter (+/- 19)
    test btree::set::difference_staggered_100_vs_10k         ... bench:      35,919 ns/iter (+/- 418)
    test btree::set::difference_staggered_10k_vs_10k         ... bench:      93,716 ns/iter (+/- 1,358)

after:

    test btree::set::difference_random_100_vs_100            ... bench:         829 ns/iter (+/- 14)
    test btree::set::difference_random_100_vs_10k            ... bench:       2,871 ns/iter (+/- 112)
    test btree::set::difference_random_10k_vs_100            ... bench:      60,580 ns/iter (+/- 675)
    test btree::set::difference_random_10k_vs_10k            ... bench:     188,386 ns/iter (+/- 1,727)
    test btree::set::difference_staggered_100_vs_100         ... bench:         863 ns/iter (+/- 21)
    test btree::set::difference_staggered_100_vs_10k         ... bench:       2,673 ns/iter (+/- 70)
    test btree::set::difference_staggered_10k_vs_10k         ... bench:      83,875 ns/iter (+/- 2,977)

Do you want me to commit it here or later? (or not at all...)

**Comment by KodrAus** [comment]
This is looking great! Thanks for giving these methods some TLC @ssomers.

> Do you want me to commit it here or later? (or not at all...)

Yeh I think we can roll difference into this PR as well while we're working on these set operations. I'm happy with the implementation of intersection now.

**Comment by ssomers** [comment]
I meant "push" instead of "commit", but github lists the commits according to time committed locally anyway. Confusing...

Anyways, it's all here now, and nothing changed to the implementation of intersection.

**Comment by ssomers** [comment]
For future reference: if you wonder why we have to tediously implement clone for Difference and for Intersection, while BTreeSet itself gets away with an easy derive(Clone), it's because of #26925. BTreeSet<T> doesn't have clone unless T has it, and that makes all the sense in the world. It doesn't make much sense for Difference and Intersection, because they hand out references to T.

**Comment by KodrAus** [comment]
Alrighty, this PR introduces some additional complexity to `BTreeSet`, but is much more efficient when performing set operations between a large and small set, which I think is a reasonable case. There's no change in the public API. So let's merge this one in!

@bors r+

**Comment by KodrAus** [comment]
@bors r-

Sorry, @ssomers do you mind squashing down some of that git history so we don't have a merge commit in there? I didn't see that one before.

**Comment by ssomers** [comment]
I don't mind, but it's going to take a while to figure out how.
